compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1986,77 +2018,104 @@ impl<'tcx> LifetimeContext<'_, 'tcx> {
-                };
-                suggest_existing(err, &name.as_str()[..], &f);
-            }
-            ([], Some("&")) if count == 1 => {
-                suggest_new(err, "&'a ");
-            }
-            ([], Some("'_")) if count == 1 => {
-                suggest_new(err, "'a");
-            }
-            ([], Some(snippet)) if !snippet.ends_with('>') => {
-                if snippet == "" {
-                    // This happens when we have `type Bar<'a> = Foo<T>` where we point at the space
-                    // before `T`. We will suggest `type Bar<'a> = Foo<'a, T>`.
-                    suggest_new(
-                        err,
-                        &std::iter::repeat("'a, ").take(count).collect::<Vec<_>>().join(""),
-                    );
-                } else {
-                    suggest_new(
-                        err,
-                        &format!(
-                            "{}<{}>",
-                            snippet,
-                            std::iter::repeat("'a").take(count).collect::<Vec<_>>().join(", ")
-                        ),
-                    );
+        match &lifetime_names[..] {
+            [name] => {
+                let mut suggs: Vec<Option<Box<dyn Fn(&str) -> String>>> = Vec::new();
+                for (snippet, count) in snippets.iter().cloned().zip(counts.iter().copied()) {
+                    if snippet == Some("&".to_string()) {
+                        suggs.push(Some(Box::new(|name| format!("&{} ", name))));
+                    } else if snippet == Some("'_".to_string()) {
+                        suggs.push(Some(Box::new(|n| n.to_string())));
+                    } else if snippet == Some("".to_string()) {
+                        suggs.push(Some(Box::new(move |n| format!("{}, ", n).repeat(count))));
+                    } else if let Some(snippet) = snippet {
+                        if !snippet.ends_with('>') {
+                            suggs.push(Some(Box::new(move |name| {
+                                format!(
+                                    "{}<{}>",
+                                    snippet,
+                                    std::iter::repeat(name.to_string())
+                                        .take(count)
+                                        .collect::<Vec<_>>()
+                                        .join(", ")
+                                )
+                            })));
+                        } else {
+                            suggs.push(None);
+                        }
+                    } else {
+                        suggs.push(None);
+                    }
+                }
+                suggest_existing(err, &name.as_str()[..], &suggs);
+            }
+            [] => {
+                let mut suggs: Vec<Option<String>> = Vec::new();
+                for (snippet, count) in snippets.iter().cloned().zip(counts.iter().copied()) {
+                    if snippet == Some("&".to_string()) {
```

Comment:
Similarly don't really like the fact that we now have to nest this. Oh well. I think a match would be cleaner still though.